src/sync/impl/sync_file.cpp

```diff
@@ -0,0 +1,262 @@
+{
+#ifdef _WIN32
+    // Not implemented yet.
+    REALM_ASSERT(false);
+#else
+    // Open the directory and list all the files.
+    DIR *dir_listing = opendir(path.c_str());
+    if (!dir_listing) {
+        return;
+    }
+    while (struct dirent *file = readdir(dir_listing)) {
+        auto file_type = file->d_type;
+        std::string file_name = file->d_name;
+        if (file_name == "." || file_name == "..") {
+            continue;
+        }
+        if (file_type == DT_REG || file_type == DT_FIFO) {
+            // Regular file (this really shouldn't ever throw, so if it does don't bother catching the exception).
+            File::remove(file_path_by_appending_component(path, file_name));
+        } else if (file_type == DT_DIR) {
+            // Directory, recurse
+            remove_nonempty_dir(file_path_by_appending_component(path, file_name, true));
+        }
+    }
+    closedir(dir_listing);
+    // Delete the directory itself
+    util::remove_dir(path);
+#endif
+}
+
+std::string make_percent_encoded_string(const std::string& raw_string)
+{
+    std::string buffer;
+    buffer.reserve(raw_string.length());
+    for (char character : raw_string) {
+        if (character_is_unreserved(character)) {
+            buffer.push_back(character);
+        } else {
+            char b[4] = {0, 0, 0, 0};
+            snprintf(b, 4, "%%%2X", character);
+            buffer.append(b);
+        }
+    }
+    return buffer;
+}
+
+std::string make_raw_string(const std::string& percent_encoded_string)
+{
+    std::string buffer;
+    size_t input_len = percent_encoded_string.length();
+    buffer.reserve(input_len);
+    size_t idx = 0;
+    while (idx < input_len) {
+        char current = percent_encoded_string[idx];
+        if (current == '%') {
+            // Decode. +3.
+            buffer.push_back(decoded_char_for(percent_encoded_string, idx));
+            idx += 3;
+        } else {
+            // No need to decode. +1.
+            if (!character_is_unreserved(current)) {
+                throw std::invalid_argument("Input string is invalid: contains reserved characters.");
+            }
+            buffer.push_back(current);
+            idx++;
+        }
+    }
+    return buffer;
+}
+
+std::string file_path_by_appending_component(const std::string& path, const std::string& component, bool is_directory)
+{
+    // FIXME: Does this have to be changed to accomodate Windows platforms?
+    std::string terminal = "";
+    if (is_directory && component[component.length() - 1] != '/') {
+        terminal = "/";
+    }
+    char path_last = path[path.length() - 1];
+    char component_first = component[0];
+    if (path_last == '/' && component_first == '/') {
+        return path + component.substr(1) + terminal;
+    } else if (path_last == '/' || component_first == '/') {
+        return path + component + terminal;
+    } else {
+        return path + "/" + component + terminal;
+    }
```

Comment:
There's a bit of heap allocation in temporaries here, consider reserving the memory up front.